### ISSUE TITLE
Ability to blacklist peers

### ISSUE THREAD

**Issue by carver** [ethereum/py-evm#893]
### What is wrong?

When disconnecting from a peer, some kinds of disconnects suggest that we should not try to reconnect to it later (is peer myself? is it replying with protocol-breaking headers? etc). It would be best to not try to connect to it again later, but nothing is preventing us from retrying.

### How can it be fixed

- Add a peer blacklist
- Check peer against blacklist before trying to connect to it
- Add peer to blacklist for certain kinds of peer failures (like a protocol error)

The blacklist can be in-memory, resetting on every `trinity` reboot.

**Comment by pipermerriam** [comment]
It'd be nice to use something with constant size storage, otherwise this will be a DOS vector.  Lets use a bloom filter and a salt to ensure that remotes cannot generate their own hash collisions.

**Comment by veox** [comment]
Two wei:

-----

> some kinds of disconnects suggest that we should not try to reconnect to it later (is peer myself? is it replying with protocol-breaking headers? etc).

A common case on the ETH public mainnet is other-genesis peers with the same network ID that otherwise follow all [sub-]protocol rules correctly.

This can get annoying/overwhelming, especially since it consumes precious resources. `geth`/`parity` APIs aren't very comprehensive on this, so in a few cases I've resorted to firewall-level connection filtering (think `iptables`/`shorewall`/`fail2ban`).

-----

> Add a peer blacklist

May I suggest more generic/neutral language?.. Like: blocklist; but really: tier, filter, sieve.

The reason is that these things tend to get layered. A not entirely fictional scenario:

* first I want to just block the irrelevant;
* but then I want a limited set of "priority" peers (e.g. a VPN ring specified in a text file);
* then there happens to be "customers" that maybe I'd drop on a rainy day, but they're paying, so I try not to;
* eventually, there's the "too slow" category that gets dropped during low load periods, and when I myself get dropped for the same reason during high load, this is the pool I'd like to dip into, instead of thrashing around randomly.

I guess the gist is: all of these involve peer management, are application-specific and can't be guessed by developers in advance.

A blacklist/blocklist is fine as a start, but long-term this is best given away to the node operator's middleware.

**Comment by carver** [comment]
> A common case on the ETH public mainnet is other-genesis peers with the same network ID that otherwise follow all [sub-]protocol rules correctly.

[We're already dropping other-genesis peers on connect](https://github.com/ethereum/py-evm/blob/master/p2p/peer.py#L522-L524). It seems reasonable to add them to the blacklist (name TBD), too.

> May I suggest more generic/neutral language?.. Like: blocklist; but really: tier, filter, sieve.
> 
> The reason is that these things tend to get layered. A not entirely fictional scenario

Yeah, we'll have to think through the best API to let people plug in their own peer discovery logic without having to redefine a bunch of basic peer handling logic. Blacklist is a simple/temporary solution.

**Comment by pipermerriam** [comment]
Relevant: https://github.com/ethereum/py-evm/issues/870

**Comment by cburgdorf** [comment]
I'm running into this as part of #945 where the transaction pool wants to black list peers that send invalid transactions.

I wanted to throw out a rough implementation outline for this:

1. Implement a new standalone service e.g. `BlockedPeerService`
2. Expose an event API that other services can consume (Not strictly speaking events but using a [generic version of our existing msg queue pattern](https://github.com/ethereum/py-evm/pull/961))
3. Events would be: `PeerAdded`, `PeerRemoved`, ...
4. API methods would be:
  - `add_peer(peer: BasePeer)`
  -  `remove_peer(peer: BasePeer)`
  -  `is_blocked(peer: BasePeer`
  - ...

Then I guess the `PeerPool` would simply listen to events from the `BlockedPeerService` to perform the actual disconnect. Services such as the `TxPool` would call `blocked_peer_service.remove_peer(douchebag)` and everyone else would receive a new event `PeerAdded` accordingly.

Thoughts?

**Comment by pipermerriam** [comment]
What do you think about making this part of the `PeerPoolSubscriber` API and allow subscribers to voice their opinion within `PeerPool.is_valid_connection_candidate`?

**Comment by cburgdorf** [comment]
>and allow subscribers to voice their opinion within `PeerPool.is_valid_connection_candidate`

Sounds like you want to maintain the blocklist **in** the `PeerPool` then?

I'm not sure if we actually need that form of coupling. Maintaining this blocklist in a separate service that exposes an API to add / check peers and raises events when new peers are added seems more flexible to me.

We would essentially keep the `PeerPool` free from any of this logic and just interact with the `BlockedPeerService` by checking `blocked_peer_service.is_blocked(peer)` as well as subscribing to the events to get notified when new peers are added to the block list to perform a disconnect.

**Comment by pipermerriam** [comment]
No, the idea is that the `PeerPool.is_valid_connection_candidate` would have a section in it which delegated checks to the subscribers.

```python
class PeerPoolSubscriber:
    def is_valid_connection_candidate(...):
        ...

class PeerPool:
    def is_valid_connection_candidate(self, ...):
        ... # current logic
        for subscriber in self.subscribers:
            if susbscriber.is_valid_connection_candidate(...):
                continue
            else:
                return False
```

Something like that (but probably all with async and probably we can call all of the subscriber methods concurrently.

**Comment by cburgdorf** [comment]
Ok, but each `PeerPoolSubscriber` doesn't want to maintain its own version of a blacklist to answer that `is_valid_connection_candidate()` call, does it?

Can you give a rundown how you anticipate some common use cases.

Here are some from my book:

**Tx Pool wants to block a peer**

1. `TxPool` recognizes `bad_peer` has fed us junk too often (using `TxPool` internal accounting)
2. `TxPool` calls `self.blocked_peer_service.block_peer(bad_peer)`
3. `BlockedPeerService` propagates `PeerBlocked` event to everyone whos listening for it
4. `PeerPool` who is among the listeners, recognizes the new blocked peer and disconnects.

**New Peer Joins the Pool**
1. A new peer joins the pool but the PeerPool checks `self.blocked_peer_service.is_blocked(new_peer)` before it lets it in and propagates the new peer to `PeerPoolSubscriber`

**Comment by carver** [comment]
My interest in this has renewed while working on syncing:

```
WARNING  08-31 14:05:18       chain  ETHPeer <Node(0x865a@52.176.100.77)> announced difficulty 9960101580041421, but didn't return any headers after <BlockHeader #913134 52277e0e>@278543309
    INFO  08-31 14:05:18        peer  ETHPeer <Node(0x865a@52.176.100.77)> finished, removing from pool
    INFO  08-31 14:05:26        peer  Adding ETHPeer <Node(0x865a@52.176.100.77)> to pool
    INFO  08-31 14:05:26       chain  Starting sync with ETHPeer <Node(0x865a@52.176.100.77)>
 WARNING  08-31 14:05:27       chain  ETHPeer <Node(0x865a@52.176.100.77)> announced difficulty 9960127846391173, but didn't return any headers after <BlockHeader #913134 52277e0e>@278543309
```

Very soon after booting a peer, they reconnect and gum up the sync (by announcing a high TD but then not returning any headers)
